core/tx_pool.go

```diff
@@ -575,22 +579,36 @@ func (pool *TxPool) add(tx *types.Transaction, local bool) (replaced bool, err e
 		knownTxMeter.Mark(1)
 		return false, ErrAlreadyKnown
 	}
+	// Make the local flag. If it's from local source or it's from the network but
+	// the sender is marked as local previously, treat it as the local transaction.
+	isLocal := local || pool.locals.containsTx(tx)
+
 	// If the transaction fails basic validation, discard it
-	if err := pool.validateTx(tx, local); err != nil {
+	if err := pool.validateTx(tx, isLocal); err != nil {
 		log.Trace("Discarding invalid transaction", "hash", hash, "err", err)
 		invalidTxMeter.Mark(1)
 		return false, err
 	}
 	// If the transaction pool is full, discard underpriced transactions
 	if uint64(pool.all.Count()) >= pool.config.GlobalSlots+pool.config.GlobalQueue {
```

Comment:
I think this check is wrong. The point of this was to check if the pool is full, and if so, drop a number of txs to make sure the new one fits. This worked well before the slot rewrite because the only way for a new tx not to fit is if the pool was exactly at capacity (or better put `pool.all.Count()+1) > pool.config.GlobalSlots+pool.config.GlobalQueue`). With the slot rewrite, I think we should have updated this part too to check `pool.all.Count()+NUMSLOTS) > pool.config.GlobalSlots+pool.config.GlobalQueue`, since if the pool has 1 slot open and we need 4, we still need to drop 3.

Reply:
The situation is not overly bad, because once we overflown by 1, it self-heals because below we had the forethought to discard based on current size, not on theoretical counts. Still, would be nice to fix the inaccuracy.

Reply:
yup